blueocean-bitbucket-pipeline/src/main/java/io/jenkins/blueocean/blueocean_bitbucket_pipeline/server/BitbucketServerEndpointContainer.java

```diff
@@ -0,0 +1,120 @@
+
+/**
+ * @author Vivek Pandey
+ */
+public class BitbucketServerEndpointContainer extends ScmServerEndpointContainer {
+    private static final Logger LOGGER = LoggerFactory.getLogger(BitbucketServerEndpointContainer.class);
+    private final Link self;
+
+    public BitbucketServerEndpointContainer(Reachable parent) {
+        this.self = parent.getLink().rel("servers");
+    }
+
+    @Override
+    public Link getLink() {
+        return self;
+    }
+
+    @Override
+    public ScmServerEndpoint create(JSONObject request) {
+        List<ErrorMessage.Error> errors = Lists.newLinkedList();
+
+        // Validate name
+        final String name = (String) request.get(ScmServerEndpoint.NAME);
+        if(StringUtils.isBlank(name)){
+            errors.add(new ErrorMessage.Error(ScmServerEndpoint.NAME, ErrorMessage.Error.ErrorCodes.MISSING.toString(), ScmServerEndpoint.NAME + " is required"));
+        }
+
+        final String url = (String) request.get(ScmServerEndpoint.API_URL);
+        if(StringUtils.isBlank(url)){
+            errors.add(new ErrorMessage.Error(ScmServerEndpoint.API_URL, ErrorMessage.Error.ErrorCodes.MISSING.toString(), ScmServerEndpoint.API_URL + " is required"));
+        }else {
+            try {
+                Response response = Request.Get(url).execute();
+                Header header = response.returnResponse().getFirstHeader("X-AREQUESTID");
+                if (header == null || StringUtils.isBlank(header.getValue())) {
+                    errors.add(new ErrorMessage.Error(BitbucketServerEndpoint.API_URL, ErrorMessage.Error.ErrorCodes.INVALID.toString(), "Specified URL is not a Bitbucket server"));
+                }
+            } catch (IOException e) {
+                errors.add(new ErrorMessage.Error(BitbucketServerEndpoint.API_URL, ErrorMessage.Error.ErrorCodes.INVALID.toString(), "Could not connect to Bitbucket server"));
+                LOGGER.error("Could not connect to Bitbucket", e);
+            }
+        }
+
+        //validate presence of endpoint with same name
+        BitbucketEndpointConfiguration endpointConfiguration = BitbucketEndpointConfiguration.get();
+
+        for(AbstractBitbucketEndpoint endpoint:endpointConfiguration.getEndpoints()){
+            if(StringUtils.isNotBlank(name)) {
+                String displayName = endpoint.getDisplayName();
+                if (StringUtils.isNotBlank(displayName) && displayName.equals(name)) {
+                    errors.add(new ErrorMessage.Error(ScmServerEndpoint.NAME, ErrorMessage.Error.ErrorCodes.ALREADY_EXISTS.toString(), ScmServerEndpoint.NAME + " already exists"));
+                }
+            }
+            if(StringUtils.isNotBlank(url)) {
+                if (url.startsWith(endpoint.getServerUrl())) {
+                    errors.add(new ErrorMessage.Error(ScmServerEndpoint.API_URL, ErrorMessage.Error.ErrorCodes.ALREADY_EXISTS.toString(), ScmServerEndpoint.API_URL + " already exists"));
+                }
+            }
```

Comment:
@stephenc  `BitbucketEndpointConfiguration.normalizeServerUrl(apiUrl)`  returns bitbucket cloud URL if apiUrl is null. Why not throw exception? Otherwise it leaves me with doing the blank/null check myself, not too bad but returning bitbucket cloud url is unexpected and stops it being from generic URL normalization utility.